x-pack/plugins/cloud/public/plugin.tsx

```diff
@@ -82,268 +63,93 @@ export interface CloudSetup {
-      if (this.isCloudEnabled) {
-        home.tutorials.setVariable('cloud', { id, baseUrl, profileUrl, deploymentUrl });
-      }
-    }
-
-    const fullCloudDeploymentUrl = getFullCloudUrl(baseUrl, deploymentUrl);
-    const fullCloudProfileUrl = getFullCloudUrl(baseUrl, profileUrl);
-    const fullCloudOrganizationUrl = getFullCloudUrl(baseUrl, organizationUrl);
-    const fullCloudSnapshotsUrl = `${fullCloudDeploymentUrl}/${CLOUD_SNAPSHOTS_PATH}`;
+    const { id, cname, base_url: baseUrl } = this.config;
 
     return {
       cloudId: id,
       cname,
       baseUrl,
-      deploymentUrl: fullCloudDeploymentUrl,
-      profileUrl: fullCloudProfileUrl,
-      organizationUrl: fullCloudOrganizationUrl,
-      snapshotsUrl: fullCloudSnapshotsUrl,
+      ...this.getCloudUrls(),
       isCloudEnabled: this.isCloudEnabled,
+      registerCloudService: (contextProvider) => {
+        this.contextProviders.push(contextProvider);
+      },
     };
   }
 
-  public start(coreStart: CoreStart, { security }: CloudStartDependencies): CloudStart {
-    const { deployment_url: deploymentUrl, base_url: baseUrl } = this.config;
+  public start(coreStart: CoreStart): CloudStart {
     coreStart.chrome.setHelpSupportUrl(ELASTIC_SUPPORT_LINK);
 
-    const setLinks = (authorized: boolean) => {
-      if (!authorized) return;
-
-      if (baseUrl && deploymentUrl) {
-        coreStart.chrome.setCustomNavLink({
-          title: i18n.translate('xpack.cloud.deploymentLinkLabel', {
-            defaultMessage: 'Manage this deployment',
-          }),
-          euiIconType: 'logoCloud',
-          href: getFullCloudUrl(baseUrl, deploymentUrl),
-        });
-      }
-
-      if (security && this.isCloudEnabled) {
-        const userMenuLinks = createUserMenuLinks(this.config);
-        security.navControlService.addUserMenuLinks(userMenuLinks);
-      }
-    };
-
-    this.checkIfAuthorizedForLinks({ http: coreStart.http, security })
-      .then(setLinks)
-      // In the event of an unexpected error, fail *open*.
-      // Cloud admin console will always perform the actual authorization checks.
-      .catch(() => setLinks(true));
-
-    // There's a risk that the request for chat config will take too much time to complete, and the provider
-    // will maintain a stale value.  To avoid this, we'll use an Observable.
+    // Nest all the registered context providers under the Cloud Services Provider.
+    // This way, plugins only need to require Cloud's context provider to have all the enriched Cloud services.
     const CloudContextProvider: FC = ({ children }) => {
-      const chatConfig = useObservable(this.chatConfig$, { enabled: false });
-      return <ServicesProvider chat={chatConfig}>{children}</ServicesProvider>;
+      return (
+        <>
+          {this.contextProviders.reduce(
+            (acc, ContextProvider) => (
+              <ContextProvider> {acc} </ContextProvider>
+            ),
+            children
+          )}
+        </>
```

Comment:
TS complains because if `this.contextProviders` is empty, it will return `children`, which may be undefined, so it does not match the `ReactNode` signature.